src/linear_algebra/pi_tensor_product.lean

```diff
@@ -0,0 +1,427 @@
+  tprod_coeff R z₁ f + tprod_coeff R z₂ f = tprod_coeff R (z₁ + z₂) f :=
+quotient.sound' $ add_con_gen.rel.of _ _ (eqv.of_add_scalar z₁ z₂ f)
+
+lemma add_tprod (f : Π i, s i) (i : ι) (m₁ m₂ : s i) :
+  (⨂ₜ[R] j, (update f i m₁) j) + (⨂ₜ[R] j, (update f i m₂) j) = ⨂ₜ[R] j, (update f i (m₁ + m₂)) j :=
+add_tprod_coeff 1 f i m₁ m₂
+
+lemma smul_tprod_coeff (z : R) (f : Π i, s i) (i : ι) (r : R) :
+  tprod_coeff R z (update f i (r • f i)) = tprod_coeff R (r * z) f :=
+quotient.sound' $ add_con_gen.rel.of _ _ $ eqv.of_smul _ _ _ _
+
+lemma smul_tprod_index (f : Π i, s i) (i j : ι) (r : R) :
+  tprod R (update f i (r • f i)) = tprod R (update f j (r • f j)) :=
+by simp_rw [tprod, smul_tprod_coeff]
+
+/-- Construct an `add_monoid_hom` from `(⨂[R] i, s i)` to some space `F` from a function
+`φ : (R × Π i, s i) → F` with the appropriate properties. -/
+def lift_add_hom (φ : (R × Π i, s i) → F)
+  (C0 : ∀ (r : R) (f : Π i, s i) (i : ι) (hf : f i = 0), φ (r, f) = 0)
+  (C0' : ∀ (f : Π i, s i), φ (0, f) = 0)
+  (C_add : ∀ (r : R) (f : Π i, s i) (i : ι) (m₁ m₂ : s i),
+    φ (r, update f i m₁) + φ (r, update f i m₂) = φ (r, update f i (m₁ + m₂)))
+  (C_add_scalar : ∀ (r r' : R) (f : Π i, s i),
+    φ (r , f) + φ (r', f) = φ (r + r', f))
+  (C_smul : ∀ (r : R) (f : Π i, s i) (i : ι) (r' : R),
+    φ (r, update f i (r' • (f i))) = φ (r' * r, f))
+: (⨂[R] i, s i) →+ F :=
+(add_con_gen (pi_tensor_product.eqv R s)).lift (free_add_monoid.lift φ) $ add_con.add_con_gen_le $
+λ x y hxy, match x, y, hxy with
+| _, _, (eqv.of_zero r' f i hf)        := (add_con.ker_rel _).2 $
+    by simp [free_add_monoid.lift_eval_of, C0 r' f i hf]
+| _, _, (eqv.of_zero_scalar f)        := (add_con.ker_rel _).2 $
+    by simp [free_add_monoid.lift_eval_of, C0']
+| _, _, (eqv.of_add z f i m₁ m₂)      := (add_con.ker_rel _).2 $
+    by simp [free_add_monoid.lift_eval_of, C_add]
+| _, _, (eqv.of_add_scalar z₁ z₂ f)      := (add_con.ker_rel _).2 $
+    by simp [free_add_monoid.lift_eval_of, C_add_scalar]
+| _, _, (eqv.of_smul z f i r')     := (add_con.ker_rel _).2 $
+    by simp [free_add_monoid.lift_eval_of, C_smul]
+| _, _, (eqv.add_comm x y)         := (add_con.ker_rel _).2 $
+    by simp_rw [add_monoid_hom.map_add, add_comm]
+end
+
+instance : has_scalar R (⨂[R] i, s i) :=
+⟨λ r, lift_add_hom (λ f : R × Π i, s i, tprod_coeff R (r * f.1) f.2)
+  (λ r' f i hf, by simp_rw [zero_tprod_coeff' _ f i hf])
+  (λ f, by simp [zero_tprod_coeff])
+  (λ r' f i m₁ m₂, by simp [add_tprod_coeff])
+  (λ r' r'' f, by simp [add_tprod_coeff', mul_add])
+  (λ z f i r', by simp [smul_tprod_coeff, (show r' * (r * z) = r * (r' * z), by ring)])⟩
+
+lemma smul_tprod_coeff' (r z : R) (f : Π i, s i) :
+  r • (tprod_coeff R z f) = tprod_coeff R (r • z) f := rfl
+
+lemma smul_tprod (r : R) (f : Π i, s i) (i : ι) :
+  r • (tprod R f) = tprod R (update f i (r • f i)) :=
+by simp [tprod, smul_tprod_coeff' r 1 f, smul_tprod_coeff]
+
+lemma smul_tprod' (r : R) (f : Π i, s i) (i : ι) (m : s i) :
```

Comment:
This lemma is true for all multilinear maps, not just tprod. It might make sense to construct `mk` earlier in the file so that you have all its projections available.

Reply:
The issue is that `mk` cannot be defined before the semimodule instance, and a lot of these small lemmas are needed to get there.

Reply:
This lemma is not used by your semimodule instance though.